src/Lean/Widget/UserWidget.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2022 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: E.W.Ayers
+-/
+import Lean.Widget.Basic
+import Lean.Data.Json
+import Lean.Environment
+import Lean.Server
+
+open Lean
+
+namespace Lean.Widget
+
+/-- A custom piece of code that is run on the editor client.
+
+The editor can use the `Lean.Widget.getWidgetSource` RPC method to
+get this object.
+
+See the [tutorial](doc/widgets.md) above this declaration for more information on
+how to use the widgets system.
+
+-/
+structure WidgetSource where
+  /-- Sourcetext of the code to run.-/
+  sourcetext : String
```

Comment:
Having a different widget id for every editor feels like the wrong approach to me, are we going to have three different copies of every widget then (in the extreme case)?  (Not that I expect this to happen. I don't think anybody is going to ship elisp or lua over the wire. But we seem to be future-proofing for this scenario.)

Reply:
The way I was thinking about it is that each object pointed to by a `widgetSourceId` is just a sourcefile that the client editor can do whatever with. Maybe in the future a `UserWidget` will point to multiple `WidgetSource` objects, eg if it depends on a library it could depend on multiple ids. So maybe `UserWidget` should have a `javascript : Hash` field in stead?

Reply:
ok I've done this